zne/noise_amplification/folding_amplifier/glorious_folding_amplifier.py

```diff
@@ -12,19 +12,30 @@
 
 """Initial Function implmention. Will be evolved and documented more in depth"""
 from qiskit.circuit.library import Barrier
+from qiskit.converters import circuit_to_dag, dag_to_circuit
 from qiskit.dagcircuit import DAGCircuit
 
 
-def apply_folding(noisy_dag: DAGCircuit, original_dag: DAGCircuit, copies: int = 2) -> DAGCircuit:
-    """Initial function for now, will be evolved and documented more in depth"""
+def _folding_to_noise_factor(folding: int) -> int:
+    """Converts number of foldings to noise factor."""
+    return 2 * folding + 1
 
-    for _ in range(copies):
-        noisy_dag.compose(original_dag, inplace=True)
-        noisy_dag.apply_operation_back(Barrier(noisy_dag.num_qubits()), qargs=noisy_dag.qubits)
-    return noisy_dag
+
+def _inverse_dag(dag_to_inverse: DAGCircuit) -> DAGCircuit:
+    """Inverts dag circuit"""
+    dag_to_inverse = dag_to_circuit(dag_to_inverse).inverse()
+    return circuit_to_dag(dag_to_inverse)
```

Comment:
This is a good utility function! I am surprised that this does not exist natively in terra. We may want to contribute it (there is a similar method called `DAGCircuit.reverse` with an implementation just like yours).

Before  doing that think about the following question: _How many times do you traverse the entire circuit with this implementation?_ (Hint: look instruction by instruction and add up)

Once you respond to that, try to respond to the same question but for a piece of code which traverses the `DAGCircuit` in reverse order (i.e. form `OutNodes` to `InNodes`) adding the inverse gates as it encounters them.

_Which one is (roughly) more efficient?_ (i.e. traverses the circuit less times and therefore is faster)

No need to change the update the code, but something to be aware of 😉